### PR TITLE
Flask session

### PR DESCRIPTION
**Pull request by bam95899**

### CODE REVIEW

File: sciencelabs/templates/macros/macros.html [hunk 1]
```diff
@@ -1,17 +1,64 @@
...
+                var select_tag = document.getElementById("semester-select");
+                var semester_id = select_tag.options[select_tag.selectedIndex].value;
+                var semester_term_year = select_tag.options[select_tag.selectedIndex].text;
+                $.ajax({
+                    type: "POST",
+                    url: "/set-semester",
+                    data: JSON.stringify({'id': semester_id, 'term-year': semester_term_year}),
+                    contentType: 'application/json;charset=UTF-8',
+                    success: function (response) {
+                        if (/month/.test(window.location.pathname))
```
Comment (c-schwarze) [inline sciencelabs/templates/macros/macros.html]: `if (/month/.test(window.location.pathname))` what is this checking?

File: sciencelabs/templates/macros/macros.html [hunk 2]
```diff
@@ -1,17 +1,64 @@
...
+                            }
+                            else if (term == 'Fall')
+                            {
+                                month = 9;
+                            }
+                            else
+                            {
+                                month = 6;
+                            }
+                            window.location.replace("/report/month/" + year + '/' + month);
```
Comment (c-schwarze) [inline sciencelabs/templates/macros/macros.html]: When you set the session semester, you don't always want to go to /reports. 🤔

File: sciencelabs/__init__.py [hunk 3]
```diff
@@ -21,7 +22,7 @@
 conn = db.connect()
 
 from sciencelabs.views import View
-from sciencelabs.session import SessionView
+from sciencelabs.sess import SessionView
```
[the remark this answers is not in the document]
Comment (bam95899) [reply sciencelabs/__init__.py]: yup, that was a naming conflict with the flask session, changed it to sessions.

File: sciencelabs/__init__.py [hunk 4]
```diff
@@ -49,6 +51,41 @@ def utility_processor():
     return to_return
 
 
+@app.before_first_request
+def create_semester_selector():
+    semester_list = Schedule().get_semesters()
+    session['SEMESTER-LIST'] = {}
+    for semester in semester_list:
+        session['SEMESTER-LIST'][(str(semester.id) + ';' + semester.term + ';' + str(semester.year))] = semester.active
```
[the remark this answers is not in the document]
Comment (bam95899) [reply sciencelabs/__init__.py]: fixed as you saw

File: sciencelabs/__init__.py [hunk 5]
```diff
@@ -49,6 +51,41 @@ def utility_processor():
...
+        return 'success'
+    else:
+        return 'error'
+
+
+# TODO IN PROGRESS LOGOUT METHOD
+@app.route("/logout", methods=["GET"])
+def logout():
+    session.clear()
+    pass
```
[the remark this answers is not in the document]
Comment (bam95899) [reply sciencelabs/__init__.py]: 👍

File: sciencelabs/sess/__init__.py [hunk 6]
```diff
@@ -54,16 +56,16 @@ def edit_session(self, session_id):
         tutor_list = self.schedule.get_registered_tutors()
         session_students = self.session.get_session_students(session_id)
         student_courses = self.session
-        semester_list = self.schedule.get_semesters()
-        course_list = self.course.get_semester_courses(40013)  # TODO: needs to update with semester selector
+        semester_list = session['SEMESTER-LIST']
+        course_list = self.course.get_semester_courses(session['SELECTED-SEMESTER'])  # TODO: needs to update with semester selector
```
[the remark this answers is not in the document]
Comment (bam95899) [reply sciencelabs/sess/__init__.py]: removed

File: sciencelabs/sess/__init__.py [hunk 7]
```diff
@@ -54,16 +56,16 @@ def edit_session(self, session_id):
...
         session_courses = self.session.get_session_courses(session_id)
         return render_template('session/edit_closed_session.html', **locals())
 
     # TODO FIX ROUTE
     @route('/attendance/edit/<int:student_id>/<int:session_id>')
     def edit_student(self, student_id, session_id):
         student = self.session.get_student_session_info(student_id, session_id)
-        student_courses = self.course.get_student_courses(student_id, 40013) #TODO: needs to update with semester selector
+        student_courses = self.course.get_student_courses(student_id, session['SELECTED-SEMESTER']) #TODO: needs to update with semester selector
```
[the remark this answers is not in the document]
Comment (bam95899) [reply sciencelabs/sess/__init__.py]: removed

File: sciencelabs/templates/macros/macros.html (hunk 2 above)
Comment (bam95899) [reply sciencelabs/templates/macros/macros.html]: This only sends you to /report/month/*year*/*month* if you are currently using the semester selector on the month report/month page. The current mathlab doesn't have the functionality to change the "active" semester while on the month page but this allows us to.

File: sciencelabs/templates/macros/macros.html (hunk 1 above)
Comment (bam95899) [reply sciencelabs/templates/macros/macros.html]: This is checking if the semester-selector was accessed on the report/month page

File: sciencelabs/db_repository/schedule_functions.py [hunk 8]
```diff
@@ -25,48 +25,66 @@ def get_yearly_schedule_tab_info(self, year, term):
...
+            .filter(Schedule_Table.deletedAt == None) \
             .filter(Session_Table.semester_id == Semester_Table.id) \
-            .filter(Semester_Table.active == 1) \
-            .group_by(Schedule_Table.id).all()
+            .filter(Semester_Table.id == semester_id) \
+            .group_by(Schedule_Table.id)\
+            .all()
 
-    def get_session_attendance(self):
+    def get_session_attendance(self, semester_id):
```
[the remark this answers is not in the document]
Comment (bam95899) [reply sciencelabs/db_repository/schedule_functions.py]: When it is passed to the method it is guaranteed to have a value.

File: sciencelabs/templates/macros/macros.html [hunk 9]
```diff
@@ -1,17 +1,64 @@
 {% macro semester_selector(semester_list) %}
     <form class="form-inline">
-        <button id="set-semester" type="button" class="btn btn-primary">Set</button>
-        <select class="custom-select">
+        <button style="border-radius: .25rem !important;" id="set-semester" type="button" class="btn btn-primary">Set</button>
```
[the remark this answers is not in the document]
Comment (bam95899) [reply sciencelabs/templates/macros/macros.html]: moved to a class

File: sciencelabs/db_repository/course_functions.py [hunk 10]
```diff
@@ -16,12 +16,12 @@ def get_course_info(self):
                 .filter(Semester_Table.active == 1)
                 .all())
 
-    def get_active_course_info(self):
+    def get_active_course_info(self, semester_id):
```
[the remark this answers is not in the document]
Comment (bam95899) [reply sciencelabs/db_repository/course_functions.py]: changed method name

File: sciencelabs/__init__.py [hunk 11]
```diff
@@ -49,6 +51,41 @@ def utility_processor():
...
+@app.route("/set-semester", methods=["POST"])
+def set_semester_selector():
+    semester_id = str(json.loads(request.data).get('id'))
+    term_year = str(json.loads(request.data).get('term-year'))
+    term_year_list = term_year.split(" ")
+    term, year = term_year_list
+    semester_string = str(semester_id) + ';' + term + ';' + str(year)
+    if semester_string in session['SEMESTER-LIST']:
+        for semester in session['SEMESTER-LIST']:
+            session['SEMESTER-LIST'][semester] = 0
```
[the remark this answers is not in the document]
Comment (bam95899) [reply sciencelabs/__init__.py]: added comments/it was refactored as you saw in my latest PR.

File: sciencelabs/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (c-schwarze) [reply sciencelabs/__init__.py]: can we name it something other than sess? Something longer than 4 letters and more descriptive? maybe `session_view`?

File: sciencelabs/templates/macros/macros.html (hunk 2 above)
Comment (c-schwarze) [reply sciencelabs/templates/macros/macros.html]: Ah. I didn't see this one in context. I like it 👍

File: sciencelabs/templates/macros/macros.html (hunk 1 above)
Comment (bam95899) [reply sciencelabs/templates/macros/macros.html]: Added a comment

File: sciencelabs/templates/macros/macros.html (hunk 2 above)
Comment (bam95899) [reply sciencelabs/templates/macros/macros.html]: 👍

**Pull request merged**